dvc/remote/local/__init__.py

```diff
@@ -581,3 +581,76 @@ def _get_unpacked_dir_names(self, checksums):
             if self.is_dir_checksum(c):
                 unpacked.add(c + self.UNPACKED_DIR_SUFFIX)
         return unpacked
+
+    def _needs_checkout(self, path_info, checksum):
+        # NOTE: In case if path_info is already cached, and cache type is
+        # 'copy' we would like to avoid relinking.
+        # `symlink/hardlink/reflink` is cheap, but if we need `copy` links,
+        # then to save on copying we should only relink if we already have
+        # `hardlink` or a `symlink`, we don't care about reflinks, because
+        # they are indistinguishable from copy anyway.
+
+        if (
+            # self._is_cache_copy(path_info)
+            not self.changed(path_info, {self.PARAM_CHECKSUM: checksum})
+            and self._is_same_link_as_cache(path_info)
+            # and not System.is_hardlink(path_info)
+            # and not System.is_symlink(path_info)
+        ):
+            msg = "File '{}' didn't change"
+            logger.debug(msg.format(str(path_info)))
+            return False
+        return True
+
+    def _is_same_link_as_cache(self, path_info):
+        is_hardlink = System.is_hardlink(path_info)
+        is_symlink = System.is_symlink(path_info)
+        is_copy_or_reflink = not is_hardlink and not is_symlink
+
+        if self.cache_types[0] == "reflink":
```

Comment:
Previously you've used is_cache_copy, which tested link types to see which one would actually work and now you are not doing it and using here first cache type in the list, which might not work in real life. E.g. by default it is `reflink, copy`, so you'll get false-positive here for reflink(for reflink it doesn't really matter, since it is the same as copy, but for hardlink/symlink it would matter).

Reply:
In default case, we will test link type, and proceed to copy, if reflink will not be working, and its ok as we consider that we cannot distinguish between reflink and copy. If cache types are ["reflink", "symlink", "hardlink"] this logic is flawed, that is true. That is why in my comment I suggested that we should be caching "working" link type so that we could build a machanism  providing proper link type and build this methods logic around it.

Reply:
We are already caching working link type, it's `self.cache_types[0]`.

Reply:
That is true, but we do not retain infromation whether it is working link type. Without context (like test or knowledge that previous linking was successufll) one cannot be sure that `self.cache_types[0]` is working link type.

Reply:
@pared So looks like we need to test the link type, right? Or did I miss it in the current PR?